src/main/java/com/theironyard/controllers/IronQuotesController.java

```diff
@@ -0,0 +1,167 @@
+            tag = new Tag(tagcommand.getValue());
+            tagRepository.save(tag);
+        }
+        quote.addTag(tag);
+        tagRepository.save(tag);
+        return tag;
+    }
+
+    @RequestMapping(path = "/quotes/{quoteId}/tags/{tagId}",method = RequestMethod.DELETE)
+    public void deleteTag(@RequestHeader(value = "Authorization") String userToken,@PathVariable Integer quoteId, @PathVariable Integer tagId){
+        getUserFromAuth(userToken);
+        Quote quote = quoteRepository.findOne(quoteId);
+        quote.deleteTag(tagRepository.getOne(tagId));
+        quoteRepository.save(quote);
+    }
+
+    @RequestMapping(path = "/quotes/{quoteId}/tags/{tagId}",method = RequestMethod.PUT)
+    public Tag updateTag(@RequestHeader(value = "Authorization") String userToken,@PathVariable Integer quoteId, @PathVariable Integer tagId ,@RequestBody TagCommand tagComannd){
+        getUserFromAuth(userToken);
+        Quote quote = quoteRepository.findOne(quoteId);
+        Tag tag = quote.editTag(tagId, tagComannd.getValue());
+        tagRepository.save(tag);
+        return tag;
+    }
+
+    @RequestMapping(path = "/users", method = RequestMethod.POST)
+    public User createUser(@RequestBody UserCommand userCommand) throws PasswordStorage.CannotPerformOperationException {
+        User user = userRepository.findByUsername(userCommand.getUsername());
+        if (user == null){
+            user = new User(userCommand.getUsername(), PasswordStorage.createHash(userCommand.getPassword()));
+            userRepository.save(user);
+        }
+        return user;
+    }
+
+    @RequestMapping(path = "/token", method = RequestMethod.POST)
+    public Map getToken(@RequestBody UserCommand userCommand) throws Exception {
+        User user = checkLogin(userCommand);
+        Map<String, String> tokens = new HashMap<>();
+        tokens.put("token", user.getToken());
+        return tokens;
+    }
+    @RequestMapping(path = "/token/regenerate",method = RequestMethod.PUT)
+    public String regenerateToken(@RequestBody UserCommand userCommand) throws Exception{
+        User user = checkLogin(userCommand);
+        return user.regenerate();
+    }
+
+    public User checkLogin(UserCommand userCommand) throws PasswordStorage.InvalidHashException, PasswordStorage.CannotPerformOperationException {
+        User user = userRepository.findByUsername(userCommand.getUsername());
+        if(user == null){
+            throw new UserNotFoundException();
+        }
+        if(!PasswordStorage.verifyPassword(userCommand.getPassword(), user.getPassword())){
+            throw new LoginFailedException();
+        }
+        if(user.getExpiration().isBefore(LocalDateTime.now())){
+            throw new TokenExpiredException();
+        }
+        return user;
+    }
+
+    public User getUserFromAuth(String userToken){
+        User user = userRepository.findFirstByToken(userToken.split(" ")[1]);
+        if(!user.isTokenValid()){
```

Comment:
This line will throw a `NullPointerException` in the event the user is not found by token.  The user needs to be checked for null and some exception thrown if the user is absent.